### PR TITLE
Dual parsimony fdr

### PR DESCRIPTION
**Pull request by zrolfs**
Primary Change:
We did and still do "Picked" protein target-decoy strategy (TDS), but the Classic TDS has been reintroduced for target proteins that were removed in the Picked TDS. These target proteins are given their classic TDS q-value, cumulative target, and cumulative decoy counts. The final protein output is sorted by q-value, so there may be some of these rescued target proteins that have much higher scores than their neighbors, but similar q-values. The Classic TDS is MORE CONSERVATIVE, so although this is not transparent, it should not be statistically concerning to include these proteins. The classic decoy proteins are not rescued for sake of output cleanliness, but exist in the q-values and cumulative decoy counts of the rescued targets.

Additional changes:
-Requiring 2+ peptides/protein was not being applied to decoys. This bug has been fixed.
-The protein q-value was calculated simply as cumulative decoy/cumulative target, and has been modified to reflect additional targets that occur at lower scores. This is how psm and peptide q-values are currently calculated. See figure 4B in [https://noble.gs.washington.edu/papers/kall2008assigning.pdf](https://noble.gs.washington.edu/papers/kall2008assigning.pdf)

### CODE REVIEW

**Comment by rmillikin** [comment]
"-Requiring 2+ peptides/protein was not being applied to decoys. This bug has been fixed."

this was intentional. we can discuss it. but it is uncommon for decoy proteins have 2 or more peptides. this leads to underestimated FDR.

**Comment by zrolfs** [comment]
The motivation for requiring two peptides per protein is to remove false-positive protein identifications. If a protein is truly a false positive, then it's unlikely to have more than one peptide. This should remove the false-positive targets with the same frequency as removing the false-positive decoys. The two peptides per protein rule can often result in fewer protein identifications because we run out of target proteins that contain 2+ peptides, but that should be the reason and not that we chose to arbitrarily favor decoys. Happy to discuss in person. Brian should be able to provide some insight on this, too.

**Comment by trishorts** [comment]
requring 2 peptides per protein inflates the number of targets because so few decoys get 2, So, you end up with too many targets above 1%. Using only one evens it out. I think this came out of Lukas Kall paper.

**Comment by zrolfs** [comment]
I don't understand how a false-positive target protein has a higher chance of getting two peptides than a false-positive decoy protein. This is an issue on the peptide level, when you inflate peptide confidence based on other peptides from the same protein, but has no impact on the protein level.

**Comment by trishorts** [comment]
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3398614/

**Comment by rmillikin** [comment]
> I don't understand how a false-positive target protein has a higher chance of getting two peptides than a false-positive decoy protein.

I think shared peptides could be one cause of this. A peptide may be present with high confidence, but its parsimony-assigned parent protein isn't, necessarily. I'm guessing that decoy-target sharing is less common than target-target sharing, but that's just a hunch.

**Comment by zrolfs** [comment]
"Contrary to the intuition, the “two-peptide” rule reduces the number of protein identifications in the target database more significantly than in the decoy database and results in **increased** false discovery rates, compared to the case when single-hit proteins are not discarded. "

**Comment by rmillikin** [comment]
The wording of the abstract is a bit confusing to me, but I do understand their/your point that apparently the two-peptide rule makes the results more conservative, not more anti-conservative. I do think this is a bit of a moot point anyway, right? Since with this PR we're not even reporting decoys, only targets?

**Comment by zrolfs** [comment]
This PR still reports decoy proteins that survived the picked TDS (so decoys that were output before this PR are still written). The point is important regardless of output, since it has a major impact on the q-value calculation and makes an already conservative method reach "don't tread on me" level of FDR overestimation.

**Comment by rmillikin** [comment]
well, I had in my mind that we'd calculate q-values not using the two-peptide rule, and then report certain proteins or not depending on whether or not the user had the two-peptide rule checkbox checked. I'd rather not have the two-peptide rule checkbox influence FDR at all... only an output difference

**Comment by zrolfs** [comment]
I hadn't thought of it that way. Whenever I check the two peptide rule in other programs, it influences the FDR, which is what I would expect. If a given protein doesn't have two peptides, I wouldn't think that it would be used for FDR calculations, since it never should have been made.

**Comment by rmillikin** [comment]
other programs I'm sure use the two-peptide rule to influence FDR... I think all of us are skeptical of it for our various reasons. I'd rather it just be a cosmetic difference but that's just my personal opinion and not the mainstream

**Comment by zrolfs** [comment]
I understand what you mean as having it be just an output filter (like only targets, or only things above 1% FDR), but because this is such a mainstream feature, I would be frustrated (as a user) if it didn't function the way I expected it to.

**Comment by blfrey** [comment]
I'm certainly no authority on this--especially not the criteria that should be used for FDR.  I lean toward the idea of using 2 peptide as a user filter only.   (One-hit wonders were semi-validated years ago; and mass specs and search strategies have improved since then).  Why should we make our lives harder just to appease old-school people who still want their 2 peptide rule to affect FDR?   That's my off-hand comment.  I'm not inculcated in the software development and user experience communities that you all are.

**Comment by trishorts** [comment]
for "requiring 2 peptides for decoys", do you mean when this box is checked? 
image

**Comment by trishorts** [comment]
hadn't digested all of the above when I made that post. But, in my mind, checking that box means that no protein gets on the list unless it has two unique peptides (including targets or decoys). Further, that no protein is used in FDR calculations unless it has two unique peptides. Is this the correct understanding?

**Comment by zrolfs** [comment]
In both approaches, no protein with <2 unique peptides is output. In this PR, no protein with <2 unique peptides is used for FDR. In the current version, proteins with <2 unique peptides ARE used in FDR.

**Comment by trishorts** [comment]
"In this PR, no protein with <2 unique peptides is used for FDR (when that box is checked)." 

and when it is not checked, I assume that only one is required?

File: EngineLayer/ProteinScoringAndFdr/ProteinScoringAndFdrEngine.cs [hunk 1]
```diff
@@ -117,62 +117,108 @@ private List<ProteinGroup> DoProteinFdr(List<ProteinGroup> proteinGroups)
...
             {
                 if (TreatModPeptidesAsDifferentPeptides)
-                    proteinGroups = proteinGroups.Where(p => p.IsDecoy || new HashSet<string>(p.AllPeptides.Select(x => x.FullSequence)).Count > 1).ToList();
+                {
+                    proteinGroups = proteinGroups.Where(p => p.AllPeptides.GroupBy(x => x.FullSequence).Count() > 1).ToList();
+                }
                 else
-                    proteinGroups = proteinGroups.Where(p => p.IsDecoy || new HashSet<string>(p.AllPeptides.Select(x => x.BaseSequence)).Count > 1).ToList();
+                {
+                    proteinGroups = proteinGroups.Where(p => p.AllPeptides.GroupBy(x => x.FullSequence).Count() > 1).ToList();
```
[the remark this answers is not in the document]
Comment (zrolfs) [reply EngineLayer/ProteinScoringAndFdr/ProteinScoringAndFdrEngine.cs]: Yep!

**Pull request merged**